src/main/java/org/elasticsearch/indices/IndicesLifecycle.java

```diff
@@ -147,6 +147,27 @@ public void afterIndexShardClosed(ShardId shardId, @Nullable IndexShard indexSha
         }
 
         /**
+         * Called before the index shard gets deleted from disk
+         * Note: this method is only executed on the first attempt of deleting the shard. Retries are will not invoke
+         * this method.
+         * @param shardId The shard id
+         * @param indexSettings the shards index settings
+         */
+        public void beforeIndexShardDeleted(ShardId shardId, Settings indexSettings) {
+        }
+
+        /**
+         * Called after the index shard has been deleted from disk.
+         *
+         * Note: this method is only called if the deletion of the shard did finish without an exception
+         *
+         * @param shardId The shard id
+         * @param indexSettings the shards index settings
+         */
+        public void afterIndexShardDeleted(ShardId shardId, Settings indexSettings) {
```

Comment:
Can you annotate this and the previous `indexSettings` with `@IndexSettings`? (I still wish we had separate classes, but it makes it easier to identify in the event this gets renamed to something non-specific).